kotlinx-coroutines-core/wasmWasi/src/internal/EventLoop.kt

```diff
@@ -0,0 +1,205 @@
+    //(ptrToSubscription + 24).storeLong(timeout)
+    //__wasi_timestamp_t precision;
+    (ptrToSubscription + 32).storeLong(0)
+    //__wasi_subclockflags_t
+    (ptrToSubscription + 40).storeShort(0) //ABSOLUTE_TIME=1/RELATIVE=0
+
+    return ptrToSubscription
+}
+
+@OptIn(UnsafeWasmMemoryApi::class)
+private fun clockTimeGet(ptrTo8Bytes: Pointer): Long {
+    val returnCode = wasiRawClockTimeGet(
+        clockId = CLOCKID_MONOTONIC,
+        precision = 1,
+        resultPtr = ptrTo8Bytes.address.toInt()
+    )
+    check(returnCode == 0)
+    return ptrTo8Bytes.loadLong()
+}
+
+@OptIn(UnsafeWasmMemoryApi::class)
+private fun sleep(timeout: Long, ptrTo32Bytes: Pointer, ptrTo8Bytes: Pointer, ptrToSubscription: Pointer) {
+    //__wasi_timestamp_t timeout;
+    (ptrToSubscription + 24).storeLong(timeout)
+
+    val returnCode = wasiPollOneOff(
+        ptrToSubscription = ptrToSubscription.address.toInt(),
+        eventPtr = ptrTo32Bytes.address.toInt(),
+        nsubscriptions = 1,
+        resultPtr = ptrTo8Bytes.address.toInt()
+    )
+
+    check(returnCode == 0)
+}
+
+@OptIn(UnsafeWasmMemoryApi::class)
+private fun sleepAndGetTime(
+    absoluteTime: Long,
+    ptrTo32Bytes: Pointer,
+    ptrTo8Bytes: Pointer,
+    ptrToSubscription: Pointer
+): Long {
+    var currentTime = clockTimeGet(ptrTo8Bytes)
+    val sleepTimeout = absoluteTime - currentTime
+
+    if (sleepTimeout > 0) {
+        sleep(
+            timeout = sleepTimeout,
+            ptrTo32Bytes = ptrTo32Bytes,
+            ptrTo8Bytes = ptrTo8Bytes,
+            ptrToSubscription = ptrToSubscription
+        )
+        currentTime += sleepTimeout
+    }
+
+    return currentTime
+}
+
+private fun runEventCycleSimple() {
+    for (currentEvent in currentCycleEvents) {
+        try {
+            currentEvent.callback?.invoke()
+        } catch (e: Throwable) {
+            thrownExceptions.add(e)
+        }
+    }
+}
+
+private fun runEventCycle(currentTime: Long) {
+    for (currentEvent in currentCycleEvents) {
+        val callback = currentEvent.callback ?: continue
+        val eventAbsoluteTime = currentEvent.absoluteTimeout
+        if (currentTime >= eventAbsoluteTime) {
+            try {
+                callback()
+            } catch (e: Throwable) {
+                thrownExceptions.add(e)
+            }
+        } else {
+            nextCycleEvents.add(currentEvent)
+            nextCycleNearestEventAbsoluteTime = min(eventAbsoluteTime, nextCycleNearestEventAbsoluteTime)
+            nextCycleContainTimedEvent = eventAbsoluteTime > 0
+        }
+    }
+}
+
+@OptIn(UnsafeWasmMemoryApi::class)
+internal fun runEventLoop() {
+    if (nextCycleEvents.isEmpty()) return
+
+    withScopedMemoryAllocator { allocator ->
+        val ptrTo32Bytes = allocator.allocate(32)
+        val ptrTo8Bytes = allocator.allocate(8)
+        val ptrToSubscription = initializeSubscriptionPtr(allocator)
+
+        while (nextCycleEvents.isNotEmpty()) {
```

Comment:
> If I understand correctly, here's how the event loop implementation works:
Is this a requirement for the implementation? Because there is no tests checking this. And I am not sure that the other implementations, like in `JS`, are following this.

Reply:
Yes, this is a requirement. Tests don't check this because, in multithreaded environments, we actually violate this intentionally sometimes. On JS, we don't use our own event loop, we rely on the one provided by the environment.

Now that I look at the implementation more closely, I don't understand the reason to implement new event loop logic at all. We can probably inherit from `EventLoopImplBase`. For that, we need to implement `EventLoopImplPlatform` for Wasm/WASI properly. If we go down that path, then, to implement `runEventLoop`, we can duplicate the logic from https://github.com/Kotlin/kotlinx.coroutines/blob/f9a4545e5d759595153dd089013cad141c96555b/kotlinx-coroutines-core/native/src/Builders.kt#L122-L144 What do you think?